### PR TITLE
TileDB co-registered attributes per band

### PR DESCRIPTION
**Pull request by normanb**
## What does this PR do?

Currently the TileDB driver supports a container of subdatasets, each dataset within this container can have a different number of bands. It is useful to be able to create this container within GDAL for TileDB, particularly so when the number of bands and the extent of associated sources line up, e,g, a DEM and an image. 

This PR adds support for registering additional datasets with a source image and to be able to write and retrieve these in an easy manner. Compared to the current subdataset support this implementation uses the band component as a third dimension which will improve access times. It is a future task to add a check to see if all the subdatasets in the `copysubdatasets` match a base raster and to implement this as a performance option in that case.

## What are related issues/pull requests?

None

## Tasklist

 - [X] Add test case(s)
 - [X] Review
 - [X] Adjust for comments
 - [X] All CI builds and checks have passed

### CODE REVIEW

File: autotest/gdrivers/tiledb_write.py [hunk 1]
```diff
@@ -86,6 +88,69 @@ def test_tiledb_write_rgb():
...
     gdaltest.tiledb_drv.Delete('tmp/tiledb_rgb')    
 
+@pytest.mark.require_driver('TileDB')
+def test_tiledb_write_attributes():
+    try:
+        import ctypes
+    except ImportError:
+        pytest.skip()
+
+    for libname in ['msvcrt', 'libc.so.6']:
```
Comment (rouault) [inline autotest/gdrivers/tiledb_write.py]: aouch ! You don't really need to go through such low level stuff to create a in-memory dataset that can be reference by a name.
```src_ds = gdal.GetDriverByName('GTiff').Create('/vsimem/temp.tif', w, h, num_bands, gdal.GDT_Int32)``` will do the job in a more compact, safer and platform independent way.
Clean up with gdal.GetDriverByName('GTiff').Delete('/vsimem/temp.tif')

File: gdal/frmts/tiledb/tiledbdataset.cpp [hunk 2]
```diff
@@ -67,6 +69,9 @@ class TileDBDataset : public GDALPamDataset
         std::unique_ptr<tiledb::ArraySchema> m_schema;
         std::unique_ptr<tiledb::FilterList> m_filterList;
 
+         char          **papszAttributes = nullptr;
```
Comment (rouault) [inline gdal/frmts/tiledb/tiledbdataset.cpp]: should be CSLDestroy()'ed

File: gdal/frmts/tiledb/tiledbdataset.cpp [hunk 3]
```diff
@@ -315,6 +319,51 @@ CPLErr TileDBRasterBand::IWriteBlock( CPL_UNUSED int nBlockXOff,
...
 
+    if ( poGDS->lpoAttributeDS.size() > 0 )
+    {
+        for ( auto const& poAttrDS: poGDS->lpoAttributeDS )
+        {
+            GDALRasterBand* poAttrBand = poAttrDS->GetRasterBand( nBand );
+            GDALDataType eAttrType = poAttrBand->GetRasterDataType();
+            int nBytes = GDALGetDataTypeSizeBytes( eAttrType );
+            int nValues = nBlockXSize * nBlockYSize;
+            void* pAttrBlock = CPLMalloc( nBytes * nValues );
```
Comment (rouault) [inline gdal/frmts/tiledb/tiledbdataset.cpp]: use VSIMalloc() and test return value

Comment (rouault) [reply gdal/frmts/tiledb/tiledbdataset.cpp]: where is pAttrBlock freed ?

File: gdal/frmts/tiledb/tiledbdataset.cpp [hunk 4]
```diff
@@ -1376,6 +1460,76 @@ TileDBDataset* TileDBDataset::CreateLL( const char *pszFilename,
...
+            std::unique_ptr<GDALDataset> poAttrDS(
+                (GDALDataset*) GDALOpen( poDS->papszAttributes[i],
+                GA_ReadOnly ));
+
+            if ( poAttrDS != nullptr )
+            {
+                const char* pszAttrName = CPLGetBasename(
+                                            poAttrDS->GetDescription());
+                // check each is co-registered
+                if ( poAttrDS->GetSpatialRef() == poDS->GetSpatialRef() )
```
Comment (rouault) [inline gdal/frmts/tiledb/tiledbdataset.cpp]: pointer equality testing will not work.  You should use poSRS1->IsSame(poSRS2) (after null-test checking of both side), or accepting both as nullptr

File: gdal/frmts/tiledb/tiledbdataset.cpp [hunk 5]
```diff
@@ -1376,6 +1460,76 @@ TileDBDataset* TileDBDataset::CreateLL( const char *pszFilename,
...
+
+        for( int i = 0;
+            poDS->papszAttributes != nullptr &&
+            poDS->papszAttributes[i] != nullptr;
+            i++ )
+        {
+            // modeling additional attributes as subdatasets
+            poDS->bHasSubDatasets = true;
+            // check each attribute is a GDAL source
+            std::unique_ptr<GDALDataset> poAttrDS(
```
Comment (rouault) [inline gdal/frmts/tiledb/tiledbdataset.cpp]: using GDALDataset::Open() will save the (GDALDataset*) cast

File: autotest/gdrivers/tiledb_write.py (hunk 1 above)
Comment (normanb) [reply autotest/gdrivers/tiledb_write.py]: :+1:

File: gdal/frmts/tiledb/tiledbdataset.cpp (hunk 2 above)
Comment (normanb) [reply gdal/frmts/tiledb/tiledbdataset.cpp]: :+1:

File: gdal/frmts/tiledb/tiledbdataset.cpp (hunk 3 above)
Comment (normanb) [reply gdal/frmts/tiledb/tiledbdataset.cpp]: my fault, I missed that. I have added `CPLFree`

File: gdal/frmts/tiledb/tiledbdataset.cpp (hunk 4 above)
Comment (normanb) [reply gdal/frmts/tiledb/tiledbdataset.cpp]: noted, thanks, didn't realize that. I am going to relax that check, it is sufficient that the width, height and number of bands is equal for an additional attribute.

File: gdal/frmts/tiledb/tiledbdataset.cpp (hunk 5 above)
Comment (normanb) [reply gdal/frmts/tiledb/tiledbdataset.cpp]: :+1:

**Comment by rouault** [comment]
A few pyflakes issues
```
autotest/gdrivers/tiledb_write.py:97:5: F841 local variable 'attr_ds' is assigned to but never used
autotest/gdrivers/tiledb_write.py:113:5: F841 local variable 'new_ds' is assigned to but never used
autotest/gdrivers/tiledb_write.py:114:5: F841 local variable 'att_ds' is assigned to but never used
autotest/gdrivers/tiledb_write.py:115:5: F841 local variable 'val_ds' is assigned to but never used
```

**Comment by rouault** [comment]
May I just ask for a better commit message, like "TileDB: add capability to define co-registered attributes per band" ?
This is to make the life of the person compiling the NEWS file easier.

**Pull request merged**